### PR TITLE
quark-goldleaf: init at 1.0.0

### PR DESCRIPTION
**Pull request by TomaSajt**
## Description of changes

This PR adds 1 package: `quark-goldleaf`
and 1 nixos module: `programs.quark-goldleaf`

This package is used for sending files between a PC and the Nintendo Switch

The actual program's name is `Quark`, however as there is already a package called `quark`.
I called this one `quark-goldleaf`, as this program was made for communicating with Goldleaf.

Link to Goldleaf's repo: https://github.com/XorTroll/Goldleaf/

The PR also adds a `udev` rule which allows the Nintendo Switch to be detected by Quark.

I found that if you disconnect and then reconnect the Nintendo Switch, only by using `MODE="0666"` will it work without rebooting the PC.

---

I saw that there have been some problems regarding deterministic builds with other maven packages.

After some research, I found that reproducible builds were first added in `maven-assembly-plugin:3.2.0` ([link](https://github.com/apache/maven-assembly-plugin/commit/d5e01847897324ba5fa2f7bfa9e025ef2f1444ed))
Also, had to set the `project.build.outputTimestamp`.
It seems like this did the job, as doing `nix build .#quark-goldleaf --rebuild` didn't complain anymore.

---

I overrode the `maven`'s `jdk` to include the nix-provided `javafx`, so that I could remove the mentions of `javafx` from `pom.xml`. This way the maven deps hash doesn't differ between platforms.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by TomaSajt** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>quark-goldleaf</li>
  </ul>
</details>

File: pkgs/applications/misc/quark-goldleaf/default.nix [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+      For the program to work properly, you will have to install Nintendo Switch udev rules.
+
+      You can either do this by enabling the NixOS module `programs.quark-goldleaf`
+      or by adding the package to `services.udev.packages`
+    '';
+    license = lib.licenses.gpl3Only;
+    mainProgram = "quark-goldleaf";
+    maintainers = with lib.maintainers; [ tomasajt ];
+    platforms = lib.platforms.unix;
```
Comment (soupglasses) [inline pkgs/applications/misc/quark-goldleaf/default.nix:94]: Has this been verified? `unix` is quite large and includes everything. BSDs, Linux, MacOS and Cygwin for Windows. Probably good to look at what platforms `JavaFX` and `udev/rules.d/` are supported on.

**Review by soupglasses — CHANGES_REQUESTED** [review]
Ultimately, this is looking good. I can definitely see the inspiration with how NS-USBLoader was packaged in #245931. I have added some comments on small changes that should be implemented.

**Comment by soupglasses** [comment]
Result of `nixpkgs-review pr 251015` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>quark-goldleaf</li>
  </ul>
</details>

Untested, was built under a binfmt enviroment.

**Comment by soupglasses** [comment]
Result of `nixpkgs-review pr 251015` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>quark-goldleaf</li>
  </ul>
</details>

**Comment by soupglasses** [comment]
Package + module is able to recognize the Nintendo Switch. Have no Goldleaf based homebrew on Switch to test its functionality fully.

File: pkgs/applications/misc/quark-goldleaf/default.nix [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    install -Dm644 target/Quark.jar $out/share/java/quark-goldleaf.jar
+
+    for size in 16 24 32 48 64 128; do
+      mkdir -p $out/share/icons/hicolor/"$size"x"$size"/apps
+      convert -resize "$size"x"$size" src/main/resources/Icon.png $out/share/icons/hicolor/"$size"x"$size"/apps/quark-goldleaf.png
+    done;
+
+    makeWrapper ${jdk'}/bin/java $out/bin/quark-goldleaf \
+        "''${gappsWrapperArgs[@]}" \
+        --prefix XDG_DATA_DIRS : "${gtk3}/share/gsettings-schemas/${gtk3.name}" \
```
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/applications/misc/quark-goldleaf/default.nix:60]: The file select dialog can't function without having access to some gsetting schemas (I think).
wrapGAppsHook should have (and did) take care of that, but I didn't want to wrap the program twice, which led to me using dontWrapGApps. However, this way, the default gsettings schemas weren't being added to the wrapper (gtk3 and gsettings-desktop-schemas). There's a section in the nixpkgs manual talking about how to make it auto-detect the schemas from buildInputs, but it wasn't working.

What I am doing here is what the schema detection was supposed to be doing, but manually adding the schema.

I saw an issue about not being able to open the file selector in ns-usbloader as well, but I haven't got to reporting it yet.

File: pkgs/applications/misc/quark-goldleaf/default.nix [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-gagIQGOiygJ0Onm0SrkbFWaovqWX2WJNx7LpSRheCLM=";
+  };
+
+  sourceRoot = "source/Quark";
+
+  patches = [
+    ./make-deterministic.patch
+    ./remove-pom-jfx.patch
+    ./put-config-in-home.patch
```
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/applications/misc/quark-goldleaf/default.nix:34]: True, I'll see what I can do about it.

File: pkgs/applications/misc/quark-goldleaf/default.nix (hunk 1 above)
Comment (TomaSajt) [reply pkgs/applications/misc/quark-goldleaf/default.nix:94]: It works on Darwin, in theory, so I'll do linux ++ darwin

File: pkgs/applications/misc/quark-goldleaf/default.nix [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  meta = {
+    inherit description;
+    changelog = "https://github.com/XorTroll/Goldleaf/releases/tag/${version}";
+    homepage = "https://github.com/XorTroll/Goldleaf#quark-and-remote-browsing";
+    longDescription = ''
+      ${description}
+
+      For the program to work properly, you will have to install Nintendo Switch udev rules.
+
+      You can either do this by enabling the NixOS module `programs.quark-goldleaf`
```
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/applications/misc/quark-goldleaf/default.nix:88]: :+1:

File: pkgs/applications/misc/quark-goldleaf/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/applications/misc/quark-goldleaf/default.nix]: I made a better patch, now placing the config in XDG_CONFIG_HOME. I'll try to upstream it.

The patch was needed, as the program would, by default, place the config file next to the jar file.

**Comment by TomaSajt** [comment]
The config-path patch has been added to upstream, will just use `fetchpatch` until the next version.

**Comment by TomaSajt** [comment]
Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>quark-goldleaf</li>
  </ul>
</details>

**Review by wegank — APPROVED** [review]

File: nixos/modules/programs/quark-goldleaf.nix [hunk 5]
```diff
@@ -0,0 +1,18 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.programs.quark-goldleaf;
+in
+{
+  options = {
+    programs.quark-goldleaf = {
+      enable = lib.mkEnableOption (lib.mdDoc "quark-goldleaf with udev rules applied");
```
Comment (wegank) [inline nixos/modules/programs/quark-goldleaf.nix:8]:
```suggestion
      enable = lib.mkEnableOption "quark-goldleaf with udev rules applied";
```

**Review by soupglasses — APPROVED** [review]

**Comment by pbsds** [comment]
Result of `nixpkgs-review pr 251015` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>quark-goldleaf</li>
  </ul>
</details>

LGTM

**Pull request merged**